src/action_state.rs

```diff
@@ -537,20 +539,140 @@ impl<A: Actionlike> Default for ActionState<A> {
-#[derive(Component, Clone, Copy, PartialEq, Eq, Hash)]
+#[derive(Component, Clone, PartialEq, Eq)]
 pub struct ActionStateDriver<A: Actionlike> {
     /// The action triggered by this entity
     pub action: A,
     /// The entity whose action state should be updated
-    pub entity: Entity,
+    pub targets: ActionStateDriverTarget,
+}
+
+/// Represents the entities that an ``ActionStateDriver`` targets.
+#[derive(Component, Clone, PartialEq, Eq)]
+pub enum ActionStateDriverTarget {
+    /// No targets
+    None,
+    /// Single target
+    Single(Entity),
+    /// Multiple targets
+    Multi(HashSet<Entity>),
+}
+
+impl ActionStateDriverTarget {
+    /// Get an iterator for the entities targeted.
+    #[inline(always)]
+    pub fn iter(&self) -> impl Iterator<Item = &Entity> {
+        match self {
+            Self::None => ActionStateDriverTargetIterator::None,
+            Self::Single(entity) => {
+                ActionStateDriverTargetIterator::Single(std::iter::once(entity))
+            }
+            Self::Multi(entities) => ActionStateDriverTargetIterator::Multi(entities.iter()),
+        }
+    }
+
+    /// Insert an entity as a target.
+    #[inline(always)]
+    pub fn insert(&mut self, entity: Entity) {
+        *self = std::mem::replace(self, Self::None).with(entity);
+    }
+
+    /// Remove an entity as a target if it's in the target set.
+    #[inline(always)]
+    pub fn remove(&mut self, entity: Entity) {
+        *self = std::mem::replace(self, Self::None).without(entity);
+    }
+
+    /// Add an entity as a target.
+    #[inline(always)]
+    pub fn add(&mut self, entities: impl Iterator<Item = Entity>) {
+        for entity in entities {
+            self.insert(entity)
+        }
+    }
+
+    /// Get the number of targets.
+    #[inline(always)]
+    pub fn len(&self) -> usize {
+        match self {
+            Self::None => 0,
+            Self::Single(_) => 1,
+            Self::Multi(targets) => targets.len(),
+        }
+    }
+
+    /// Returns true if there are no targets.
+    #[inline(always)]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Add an entity as a target using a builder style pattern.
+    #[inline(always)]
+    pub fn with(mut self, entity: Entity) -> Self {
+        match self {
+            Self::None => Self::Single(entity),
+            Self::Single(og) => Self::Multi(HashSet::from([og, entity])),
+            Self::Multi(ref mut targets) => {
+                targets.insert(entity);
+                self
+            }
+        }
+    }
+
+    /// Remove an entity as a target if it's in the set using a builder style pattern.
+    pub fn without(mut self, entity: Entity) -> Self {
+        match self {
+            Self::None => Self::None,
+            Self::Single(_) => Self::None,
+            Self::Multi(ref mut targets) => {
+                targets.remove(&entity);
+                match targets.len() {
```

Comment:
This is probably nicer if you just impl `FromIter` and then use that trait here.